feat: Add doc flag and listtemplate command

jhorst11 opened the pull request:
This PR does the following:

- adds a `--doc` flag to the `listtemplates` command which will output additional descriptive documentation from the template's (optional) `template.json` file
- adds a new `listtemplate` command which shows details about a single template
- adds a `--doc` flag to the new `listtemplate` command which will output verbose documentation about the template

Example usages:

1) adding `--doc` to existing `listtemplates` command

```bash
➜  SalesforceMobileSDK-Package git:(docs-enhanced) ./ios/forceios.js listtemplates --templatesource=/Users/justin.horst/Repos/SalesforceMobileSDK-Templates --doc                           

Available templates from custom repository:

Repository: /Users/justin.horst/Repos/SalesforceMobileSDK-Templates
1) Swift application using MobileSync, SwiftUI and Combine
forceios createwithtemplate --templatesource=/Users/justin.horst/Repos/SalesforceMobileSDK-Templates --template=iOSNativeSwiftTemplate
   Description: A basic iOS native Swift application template using Salesforce Mobile SDK with MobileSync, SwiftUI, and Combine. This template provides a foundation for building iOS apps that integrate with Salesforce, featuring basic authentication, account listing, and contact management capabilities.
   Use Case: Use this template when you need to create a basic iOS native app that integrates with Salesforce using Swift, SwiftUI, and Combine. It's ideal for developers starting with Salesforce Mobile SDK who want a minimal but functional starting point.
   Features: native-ios, swift, swiftui, combine, salesforce-sdk, mobilesync, authentication, account-list, contact-management, smartstore, offline-capability
   Complexity: simple
2) Sample Swift application using MobileSync data framework
forceios createwithtemplate --templatesource=/Users/justin.horst/Repos/SalesforceMobileSDK-Templates --template=MobileSyncExplorerSwift
   Description: A comprehensive iOS Swift application template demonstrating advanced Salesforce Mobile SDK integration using MobileSync data framework. This template provides a full-featured contact management app with offline sync capabilities, search functionality, and a modern SwiftUI interface with split-view navigation.
   Use Case: Use this template when you need to create a production-ready iOS app with full Salesforce data synchronization capabilities. It's ideal for developers building apps that require offline functionality, complex data relationships, and professional-grade user interfaces.
   Features: native-ios, swift, swiftui, salesforce-sdk, mobilesync, offline-sync, smartstore, contact-management, search-functionality, split-view-navigation, notifications, settings-management, widget-support, dark-mode-support
   Complexity: moderate
```
   
2) new `listtemplate` command (no `--doc` flag)
   
 ``` bash
 ➜  SalesforceMobileSDK-Package git:(docs-enhanced) ./ios/forceios.js listtemplate --template=iOSNativeSwiftTemplate                                              

Template from default repository:

Template: iOSNativeSwiftTemplate
Description: Swift application using MobileSync, SwiftUI and Combine
App Type: native_swift
Platforms: ios

Usage:
forceios createwithtemplate --templatesource=https://github.com/forcedotcom/SalesforceMobileSDK-Templates#dev --template=iOSNativeSwiftTemplate
 ```
 
 3) new `listtemplate` command (with `--doc` flag)
 
 ```bash
   SalesforceMobileSDK-Package git:(docs-enhanced) ./ios/forceios.js listtemplate --template=iOSNativeSwiftTemplate --templatesource=/Users/justin.horst/Repos/SalesforceMobileSDK-Templates --doc

Template from custom repository:

Repository: /Users/justin.horst/Repos/SalesforceMobileSDK-Templates
Template: iOSNativeSwiftTemplate
Description: Swift application using MobileSync, SwiftUI and Combine
App Type: native_swift
Platforms: ios

Usage:
forceios createwithtemplate --templatesource=/Users/justin.horst/Repos/SalesforceMobileSDK-Templates --template=iOSNativeSwiftTemplate

Template Metadata (template.json):
{
  "$schema": "../template-schema.json",
  "templateId": "ios-native-swift-template",
  "version": "0.0.1",
  "platform": "ios",
  "displayName": "iOS Native Swift Template",
  "description": "A basic iOS native Swift application template using Salesforce Mobile SDK with MobileSync, SwiftUI, and Combine. This template provides a foundation for building iOS apps that integrate with Salesforce, featuring basic authentication, account listing, and contact management capabilities.",
  "useCase": "Use this template when you need to create a basic iOS native app that integrates with Salesforce using Swift, SwiftUI, and Combine. It's ideal for developers starting with Salesforce Mobile SDK who want a minimal but functional starting point.",
  "features": [
    "native-ios",
    "swift",
    "swiftui",
    "combine",
    "salesforce-sdk",
    "mobilesync",
    "authentication",
    "account-list",
    "contact-management",
    "smartstore",
    "offline-capability"
  ],
  "complexity": "simple",
  "customizationPoints": [
    {
      "feature": "app-configuration",
      "description": "Customize the app name, package name, and organization details",
      "files": [
        "template.js",
        "package.json",
        "iOSNativeSwiftTemplate.xcodeproj/project.pbxproj",
        "iOSNativeSwiftTemplate/iOSNativeSwiftTemplate.entitlements"
      ],
      "instructions": [
        "Update template.js with new app name, package name, and organization",
        "Modify package.json with new app name",
        "Update Xcode project files with new identifiers",
        "Update entitlements file with new package name"
      ]
    },
    {
      "feature": "salesforce-objects",
      "description": "Add new Salesforce objects beyond the default Account and Contact",
      "files": [
        "iOSNativeSwiftTemplate/AccountsListModel.swift",
        "iOSNativeSwiftTemplate/ContactsForAccountModel.swift",
        "iOSNativeSwiftTemplate/ContactDetailModel.swift"
      ],
      "instructions": [
        "Create new model files for additional SObjects",
        "Implement ObservableObject pattern with @Published properties",
        "Add SmartStore queries for new objects",
        "Create corresponding SwiftUI views for new objects"
      ]
    },
    {
      "feature": "authentication-flow",
      "description": "Customize the authentication and login experience",
      "files": [
        "iOSNativeSwiftTemplate/AppDelegate.swift",
        "iOSNativeSwiftTemplate/LoginTypeSelectionViewController.swift"
      ],
      "instructions": [
        "Modify AppDelegate.swift to customize SDK initialization",
        "Update LoginTypeSelectionViewController for custom login flows",
        "Configure push notification handling if needed",
        "Add custom authentication logic"
      ]
    },
    {
      "feature": "ui-components",
      "description": "Extend the SwiftUI interface with additional components",
      "files": [
        "iOSNativeSwiftTemplate/AccountsListView.swift",
        "iOSNativeSwiftTemplate/ContactsForAccountListView.swift",
        "iOSNativeSwiftTemplate/ContactDetailsView.swift"
      ],
      "instructions": [
        "Add new SwiftUI views for additional functionality",
        "Implement navigation between new views",
        "Add search and filtering capabilities",
        "Customize the UI theme and styling"
      ]
    }
  ]
}
 ```

jhorst11 committed:
add doc flag to listtemplates to get more verbose documentation

jhorst11 committed:
Refactor listTemplates to use shared display function and update template metadata extraction from template.json

jhorst11 committed:
Merge remote-tracking branch 'origin/dev' into docs-enhanced

jhorst11 committed:
Add listtemplate command to retrieve and display details for a specific Mobile SDK template

- Introduced `listTemplate` function in `configHelper.js` to handle template detail retrieval.
- Updated `constants.js` to include `listtemplate` command.
- Enhanced `oclifAdapter.js` to support the new command and integrate with existing template display functions.
- Implemented `getTemplate` and `displayTemplateDetail` in `templateHelper.js` for detailed template information display.

wmathurin reviewed (commented):
LGTM!

Maybe we could call the command something like 'describetemplate' instead of 'listtemplate' ?

jhorst11 committed:
Enhance template commands to support JSON output format

- Added `json` argument to `listtemplates` and `listtemplate` commands in `constants.js` for JSON output.
- Updated `listTemplates` and `listTemplate` functions in `oclifAdapter.js` to handle the new `json` argument.
- Modified `displayTemplateList` and `displayTemplateDetail` in `templateHelper.js` to output template information in JSON format when the `json` flag is specified.
- Adjusted `configHelper.js` to parse the `json` argument and pass it to the display functions.

wmathurin reviewed (commented):
LGTM overall - @jhorst11 why do you think about calling listtemplate something like describetemplate since it does not return a list, but returns details about a single template?

jhorst11 commented:
> LGTM overall - @jhorst11 why do you think about calling listtemplate something like describetemplate since it does not return a list, but returns details about a single template?

that's fine with me - during design I had it named `doc` but @khawkins recommended `listtemplate` for consistency with other command names - @khawkins do you have strong feelings about renaming to `describetemplate`?

jhorst11 committed:
refactor: rename listtemplate command to describetemplate for better clarity

khawkins commented:
> @khawkins do you have strong feelings about renaming to describetemplate?

@jhorst11 I do not. With `doc` I mostly worried that it was a bit too ambiguous. I'm okay with `describetemplate`.

jhorst11 commented:
Thanks! @wmathurin given @khawkins comment this is good to merge
